src/lasfilter_voxelgrid.cpp

```diff
@@ -0,0 +1,30 @@
+#include "lasfilter_voxelgrid.hpp"
+
+void voxelGrid::setVoxel(double vLen){
+    voxelSideLength = vLen;
+}
+
+int voxelGrid::getLength(double first, double last){
+    double distance = last - first;
+    int n = floor(distance / voxelSideLength);
+    return n;
+}
+
+bool voxelGrid::checkRegistry(double x, double y, double z){
+
+    int nx = getLength(0, x);
+    int ny = getLength(0, y);
+    int nz = getLength(0, z);
+
+    std::vector<int> nVals = {nx, ny, nz};
+
+    int nReg = dynamic_registry.size();
+    dynamic_registry.insert(nVals);
```

Comment:
It does not seems memory efficient to store a `std::vector<int>` of size 3 to test if the voxel is already filled. What about computing the voxel number/ID into an `unsigned int` and store only this id?

Reply:
That's the tricky part... on a former version of the algorithm I established minimum coordinates for the voxel registry based on the cloud's bounding box (as I mentioned in my first comment), this made indexation easier and allowed the usage of `unsigned int` as indexers.

In TLS and MLS is common practice to work in local coordinates, this means that often times the point clouds have negative coordinates on all axes.

In the current implementation, negative values are assigned negative indices, which is fine since we are filling and measuring a `set` object. One alternative would be to use very low negative values (e.g. -9999) instead of zero as a registry reference (instead of the 0 we discussed in your first reviewed piece), this way it would be safe to assume we'll always have positive indices in the registry.

And about putting the *voxel number/ID into an `unsigned int`*, it's a great idea, I'm just not sure how to approach this issue so far, without any *a priori* information from the bounding box, for example.

`-thin_with_grid` assigns values to cells on streaming, not taking any prior information to assemple a registry or anything... it's all done on the fly and many aspects of that code I cannot quite grasp yet in order to be more memory efficient.

Reply:
> I'm just not sure how to approach this issue so far, without any a priori information from the bounding box, for example

Right this is the biggest issue!

>  it's all done on the fly and many aspects of that code I cannot quite grasp yet in order to be more memory efficient

I read the code. Not easy indeed. I'm not able to understand how it manages the computation without prior knowledge on the bounding box.